mongo_connector/dest_mapping.py

```diff
@@ -15,143 +15,230 @@
+        """A utility function to set the corresponding plain variables"""
+        # The lock is necessary to properly check that multiple namespaces
+        # into a single namespace.
+        with self.lock:
+            src_name = mapped_namespace.source_name
+            target_name = mapped_namespace.dest_name
+            existing_src = self.reverse_plain.get(target_name)
+            if existing_src and existing_src != src_name:
+                raise errors.InvalidConfiguration(
+                    "Multiple namespaces cannot be combined into one target "
+                    "namespace. Trying to map '%s' to '%s' but there already "
+                    "exists a mapping from '%s' to '%s'" %
+                    (src_name, target_name, existing_src, target_name))
 
-    def replace(self, src_match, map_pattern):
-        """Given the matchobject src_match,
-        replace corresponding '*' in map_pattern with src_match."""
-        wildcard_matched = src_match.group(1)
-        return map_pattern.replace("*", wildcard_matched)
+            self.plain[src_name] = mapped_namespace
+            self.reverse_plain[target_name] = src_name
+            src_db, _ = src_name.split(".", 1)
+            target_db, _ = target_name.split(".", 1)
+            self.plain_db.setdefault(src_db, set()).add(target_db)
 
-    def get(self, plain_src_ns, defaultval=""):
-        """Given a plain source namespace, return a mapped namespace if matched.
-        If no match and given defaultval, return it as a default value.
+    def get(self, plain_src_ns):
+        """Given a plain source namespace, return a mapped namespace if it
+        should be included or None.
         """
-        with self.lock:
-            # search in plain mappings first
-            try:
-                return self.plain[plain_src_ns]
-            except KeyError:
-                # search in wildcard mappings
-                # if matched, get a replaced mapped namespace
-                # and add to the plain mappings
-                for k, v in self.wildcard.items():
-                    m_res = self.match(plain_src_ns, k)
-                    if m_res:
-                        res = self.replace(m_res, v)
-                        if res is not None:
-                            self.set_plain(plain_src_ns, res)
-                            return res
-
-            if defaultval:
-                return defaultval
-            else:
-                LOG.warn("Failed to find matched mapping "
-                         "for %s." % plain_src_ns)
-                return None
+        # if plain_src_ns matches ex_namespace_set, ignore
+        if plain_src_ns in self.ex_namespace_set:
+            return None
+        if not self.regex_map and not self.plain:
+            # here we include all namespaces
+            return MappedNamespace(plain_src_ns)
+        # search in plain mappings first
+        try:
+            return self.plain[plain_src_ns]
```

Comment:
Does this need to be protected with the lock? It seems that we hold the lock elsewhere when we read/modify `self.plain`.